nixos/nufetch: init module

gignsky opened the pull request:
<!--
This change does two things:
1. creates a package called nufetch-for-nixos-module that is simply a version of neofetch patched with a change to the default config directory so that the default config can be located in /etc rather than the user's home directory, this allows for a system-wide installation of neofetch
2. creates a nixos-module under programs/ called nufetch that allows for neofetch options to be declared via nix; it also wraps neofetch with a new command called `nufetch` that will result in the same output as `neofetch`
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [X] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

gignsky committed:
maintainers: add gignsky

gignsky committed:
nufetch-for-nixos-module: added package nececcary for nixos-module

This package is nececcary for the module to be installed.

This patch simply changes neofetch's default config directory to
/etc/nufetch.conf rather than the default home based config directory.

This is needed so that the nixos-module can put the system-wide config
file at /etc/nufetch.conf rather than building in the home directory at
the default ~/.config/neofetch/config.conf

gignsky committed:
nixos/nufetch: init module: added to module-list.nix

Added module to module-list.nix so that it can be imported

gignsky committed:
nixos/nufetch: init module: added module details

gignsky committed:
nixos/nufetch: init module: added test to all-tests.nix

gignsky committed:
nixos/nufetch: init module: added test for module

gignsky committed:
nufetch-for-nixos-module: added passthru test

gignsky committed:
nixos/nufetch: init module: added reference to 25.11 manual

Added new addition of nixos module to the appropriate section of the
25.11 manual :)

gignsky committed:
maintainers: add gignsky to neofetch

added myself as a maintainer for neofetch as it is a upstream dependancy
of my current nufetch project

gignsky committed:
formatting: output of `nix fmt` command

ran nix fmt on all files and these are the changes it outputted

gignsky committed:
nufetch-for-nixos-module: updated rev and sha256

gignsky committed:
nufetch-for-nixos-module: added patch file

This patch file is neccecary for changing neofetch's default config
directory

gignsky committed:
nufetch-for-nixos-module: fixed to not rely on exterior import

this was previously relying on an import from my github repo and has
been changed to apply the patch all within nixpkgs

pluiedev commented on nixos/modules/programs/nufetch/nufetch.nix:
```diff
@@ -0,0 +1,180 @@
+  ...
+}:
+
+with lib;
```
New uses of `with lib;` should be avoided as the overly wide scope makes it unclear where variables are coming from

pluiedev commented on pkgs/by-name/nu/nufetch-for-nixos-module/package.nix:
```diff
@@ -0,0 +1,25 @@
+pkgs.neofetch.overrideAttrs (old: {
+  pname = "nufetch-for-nixos-module";
+  patches = (old.patches or [ ]) ++ [ ./patches/neofetch-nixos-module.patch ];
+  postPatch =
+    (old.postPatch or "")
+    + ''
+      echo ">>> neofetch-nixos-module.patch applied"
+    '';
```
```suggestion
```

pluiedev commented on pkgs/by-name/nu/nufetch-for-nixos-module/package.nix:
```diff
@@ -0,0 +1,25 @@
+{
```
There's frankly no reason why this should be in by-name and visible to everyone. The derivation should be located in or near the module since by itself it scantly makes sense. The patch could be entirely eliminated too that way, since then you're just making a wrapper

gignsky replied:
I'm moving this into the module directory, but I'm not sure how you mean I could remove the patch file because while I am wrapping the program I have to patch its source to change the neofetch config directory path inside the bash script

pluiedev reviewed (changes requested):
Hi - overall I'm very confused as to the point of this PR. I'm aware that this is an attempt to upstream an existing Nufetch project, but shouldn't this be instead incorporated an actual Neofetch module? (Also - Neofetch is not even maintained anymore, so is there any point to this? Why not Fastfetch?)

gignsky commented:
> Hi - overall I'm very confused as to the point of this PR. I'm aware that this is an attempt to upstream an existing Nufetch project, but shouldn't this be instead incorporated an actual Neofetch module? (Also - Neofetch is not even maintained anymore, so is there any point to this? Why not Fastfetch?)

Hey there, this is my first PR into nixpkgs so forgive me if I ask some dumb questions. You are correct that this is an attempt to upstream github.com/gignsky/nufetch (a flake I wrote to solve an issue another user was having while trying to bash together a neofetch config with nix), I think that is what you mean? 

What do you mean by incorporating this into an actual neofetch module? Are you suggesting that rather than call this nufetch it should just be a module called neofetch? If that is the better approach I am happy to change the names around so that this is the case!

As for neofetch no longer being maintained, while this is true many users including myself still use it and I can imagine that having the ability to extend it via nix would be helpful to more people than just myself. I was also not aware fastfetch existed but will certainly look into it, maybe it needs a module too if it is extensible in a similar way to neofetch, I'll look into that after we get this PR resolved.

Looking into your suggested changes now...

gignsky committed:
nixos/nufetch: suggested update to config-maker.nix

Co-authored-by: Leah Amelia Chen <github@acc.pluie.me>

gignsky committed:
nixos/nufetch: suggested changes to config-maker.nix #2

Co-authored-by: Leah Amelia Chen <github@acc.pluie.me>

gignsky committed:
nixos/nufetch: followed suggestion to remove with lib;

gignsky committed:
nixos/nufetch: followed suggestion to use lib.getExe

Co-authored-by: Leah Amelia Chen <github@acc.pluie.me>

gignsky commented on nixos/modules/programs/nufetch/nufetch.nix:
```diff
@@ -0,0 +1,180 @@
+      defaultText = "pkgs.nufetch-for-nixos-module";
+      description = "Set version of nufetch package to use.";
+    };
+    os = mkOption {
```
I took a look at RFC 42, I don't know how this could work with neofetch's config file as the user is not always setting things to key value pairs but might be putting in their own special code that makes the string method of passing extra args better. Maybe I am misunderstanding but my thoughts are that the extraArgs should be passed as a long string

gignsky committed:
nixos/nufetch: changed to use the mkEnableOption with // { default = true; } when needed

gignsky committed:
nufetch-for-nixos-module: followed suggestion to remove postPatch debug line

Co-authored-by: Leah Amelia Chen <github@acc.pluie.me>

gignsky committed:
nixos/nufetch: added lib to maintainers

gignsky committed:
nixos/nufetch: added lib to type call

gignsky committed:
nixos/nufetch: changed extraFields back to mkOption rather than mkEnableOption

gignsky committed:
nixos/nufetch: fixed syntax bugs

gignsky committed:
nixos/nufetch: reverted away from mkEnableOption

Could not figure out how to change the mkOption args to mkEnableOption
and still have defaults, these were erroring out when trying to run
tests and build. Any help getting these working would be appreciated! :)

gignsky committed:
nixos/nufetch: cleaned up tests

gignsky committed:
nixos/nufetch: moved patched neofetch package to module path

gignsky committed:
nixos/nufetch: fixed nufetch patched version in module

gignsky committed:
formatter: ran nix fmt

gignsky commented:
> 
> > Hey there, this is my first PR into nixpkgs so forgive me if I ask some dumb questions. 
> 
> No worries!
> 
> > You are correct that this is an attempt to upstream github.com/gignsky/nufetch (a flake I wrote to solve an issue another user was having while trying to bash together a neofetch config with nix), I think that is what you mean? 
> 
> Yes.
> 
> > What do you mean by incorporating this into an actual neofetch module? Are you suggesting that rather than call this nufetch it should just be a module called neofetch? 
> 
> I think so, yeah. 
> 
> > As for neofetch no longer being maintained, while this is true many users including myself still use it and I can imagine that having the ability to extend it via nix would be helpful to more people than just myself. 
> 
> In Nixpkgs we try not to add new functionality to or based on software that's already abandoned upstream - Neofetch is somewhat of an exception because of how popular it still is. 
> 
> > I was also not aware fastfetch existed but will certainly look into it, maybe it needs a module too if it is extensible in a similar way to neofetch, I'll look into that after we get this PR resolved.
> 
> Fastfetch as far as I know uses a [JSON (with comments) config file](https://github.com/fastfetch-cli/fastfetch/wiki/Configuration) instead of a Bash script, which makes sense since it's a native C program and not a Bash script. There isn't a module for it in NixOS at the moment, though it would be a pretty standard [RFC 42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md)-style module in my eyes. 
> 
> Overall IMO an approach like this (generating a Bash script) is fairly unconventional in NixOS and it's also really hard to maintain, so personally I wouldn't consider this to be something I want to upstream - both because once it's included in NixOS its API cannot contain any breaking changes for at least one release cycle, and especially since Neofetch isn't even maintained anymore and is up for removal whenever the costs of maintaining it in Nixpkgs outweighs its benefits. Leaving this as an opinionated third-party flake is the best option here, I think, since not everyone would be happy about how this API is laid out.
> 
> 

I appreciate all your insight, I'm going to leave this as a flake and close the PR following your suggestion. I have some time coming up, I will look into a adding a module for fastfetch when I get a moment, I think that would be a great edition! 

Thank you so much for all your help! See ya in the cyberverse!
